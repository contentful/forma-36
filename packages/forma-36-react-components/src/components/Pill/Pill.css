@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/borders';

.Pill {
  display: inline-flex;
  font-family: var(--font-stack-primary);
  border-radius: var(--border-radius-small);
  border: none;
  padding: 0;
  max-width: 100%;
  composes: focus-outline--default from './../../styles/settings/a11y.css';
}

.Pill--deleted {
  background: var(--color-element-light);
  text-decoration: line-through;
  opacity: 0.5;
}

.Pill--idle {
  background: var(--color-element-light);
}

.Pill--active {
  background: var(--color-element-dark);
}

.Pill__label {
  line-height: 1.5rem;
  padding: 0.375rem 0.625rem;
  font-size: 0.875rem;
  flex-grow: 2;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.Pill__label--deleted,
.Pill__label--idle {
  color: var(--color-text-base);
}

.Pill__label--active {
  color: var(--color-text-dark);
}

.Pill__close-button {
  border: none;
  border-top-right-radius: var(--border-radius-small);
  border-bottom-right-radius: var(--border-radius-small);
  padding: 0.375rem;
  background: transparent;
  transition: background var(--transition-duration-short)
    var(--transition-easing-default);
  composes: focus-outline--default from './../../styles/settings/a11y.css';

  &:hover,
  &:focus {
    background: var(--color-element-mid);
    cursor: pointer;
  }
}

.Pill__close-button--deleted,
.Pill__close-button--idle {
  border-left: 1px solid var(--color-element-darkest);
}

.Pill__close-button--active {
  border-left: 1px solid var(--color-text-lightest);
  &:hover,
  &:focus {
    background: var(--color-element-darkest);
    cursor: pointer;
  }
}

.Pill__drag-icon {
  line-height: 1.3rem;
  padding: 0.375rem 0.625rem;
  padding-right: 0;
}

.Pill__icon--deleted,
.Pill__icon--idle {
  fill: var(--color-element-darkest);
  vertical-align: middle;
}

.Pill__icon--active {
  fill: var(--color-text-lightest);
  vertical-align: middle;
}
