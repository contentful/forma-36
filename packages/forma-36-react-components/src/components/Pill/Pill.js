import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import Icon from '../Icon';
import { iconName } from '../Icon/constants';
import TabFocusTrap from '../TabFocusTrap';
import styles from './Pill.css';

class Pill extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    testId: PropTypes.string,
    label: PropTypes.string.isRequired,
    onClose: PropTypes.func,
    onDrag: PropTypes.func,
  };

  static defaultProps = {
    extraClassNames: undefined,
    onDrag: undefined,
    testId: 'cf-ui-pill',
    onClose: undefined,
  };

  render() {
    const {
      label,
      onClose,
      testId,
      onDrag,
      extraClassNames,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.Pill, extraClassNames);

    return (
      <div
        className={classNames}
        {...otherProps}
        draggable={!!onDrag}
        onDrag={onDrag}
      >
        <TabFocusTrap>
          <span className={styles.Pill__label}>
            {onDrag && (
              <span className={styles['Pill__drag-icon']}>
                <Icon
                  icon={iconName.Drag}
                  color="muted"
                  extraClassNames={styles.Pill__icon}
                />
              </span>
            )}
            {label}
          </span>
          {onClose && (
            <button
              type="button"
              onClick={onClose}
              className={styles['Pill__close-button']}
              data-test-id={testId}
            >
              <TabFocusTrap>
                <Icon
                  icon={iconName.Close}
                  color="muted"
                  extraClassNames={styles.Pill__icon}
                />
              </TabFocusTrap>
            </button>
          )}
        </TabFocusTrap>
      </div>
    );
  }
}

export default Pill;
