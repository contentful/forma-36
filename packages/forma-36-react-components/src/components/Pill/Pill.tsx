import React from 'react';
import cn from 'classnames';
import Icon from '../Icon/Icon';
import TabFocusTrap from '../TabFocusTrap';

import styles from './Pill.css';

export interface PillProps {
  label: string;
  onClose?: () => void;
  onDrag?: () => void;
  className?: string;
  testId?: string;
  style?: React.CSSProperties;
  dragHandleComponent?: React.ReactNode;
}

export function Pill({
  label,
  onClose,
  testId,
  onDrag,
  className,
  dragHandleComponent,
  ...otherProps
}: PillProps): React.ReactElement {
  const classNames = cn(styles.Pill, className);

  return (
    <div
      className={classNames}
      data-test-id={testId}
      {...otherProps}
      onDrag={onDrag}
    >
      {onDrag &&
        (dragHandleComponent ? (
          dragHandleComponent
        ) : (
          <span className={styles['Pill__drag-icon']}>
            <Icon icon="Drag" color="muted" className={styles.Pill__icon} />
          </span>
        ))}
      <span aria-label={label} title={label} className={styles.Pill__label}>
        {label}
      </span>
      {onClose && (
        <button
          type="button"
          aria-label="close"
          onClick={onClose}
          className={styles['Pill__close-button']}
        >
          <TabFocusTrap>
            <Icon icon="Close" color="muted" className={styles.Pill__icon} />
          </TabFocusTrap>
        </button>
      )}
    </div>
  );
}

Pill.defaultProps = {
  testId: 'cf-ui-pill',
};

export default Pill;
