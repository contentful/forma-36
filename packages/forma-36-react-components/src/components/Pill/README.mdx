---
title: 'Pill'
type: 'component'
status: 'stable'
slug: /components/pill/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/Pill'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-pill--default'
---

Pills represent a set of one or multiple objects which can be entered or changed by the user. By default, pills can be removed by the user. When necessary, the order of pills can be changed by the user using drag and drop. They are used to represent complex entities in small blocks such as contact details. Contentful Pills are are used to render transition locales and tag elements.

## Table of contents

- [How to use Pill](#how-to-use-pill)
- [Code examples](#code-examples)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use Pill

- use Pill to represent data like contact details or tagging
- it can be used for web content that needs to be labeled, categorized, or organized using keywords that describe them.

## Code examples

Example of usage with draggable option

```jsx
// import { Pill } from '@contentful/forma-36-react-components';

<Pill
  tabIndex={0}
  testId="pill-item"
  label="test label"
  onClose={() => {}}
  onDrag={() => {}}
/>
```

Example of the Pill component with dragHandle, in that case Icon.

```jsx
// import { Pill, Icon } from '@contentful/forma-36-react-components';

<Pill
  label="example.user@contentful.com"
  dragHandleComponent={
    <Icon
      icon="ArrowUp"
      color="positive"
      style={{ padding: '0.375rem 0.625rem' }}
    />
  }
/>
```

## Content recommendations:

- Be sure the pill text is clear and concise.

## Accessibility

- Label value is passed in the component as a aria-label and title in the span element, like follows `aria-label={label} title={label}`.
