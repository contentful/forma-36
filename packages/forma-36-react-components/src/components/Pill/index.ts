export { Pill } from './Pill';
export type { PillProps } from './Pill';
