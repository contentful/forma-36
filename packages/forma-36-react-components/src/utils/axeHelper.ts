import { configureAxe } from 'jest-axe';

const axe = configureAxe({
  rules: {
    region: { enabled: false },
  },
});

export default axe;
