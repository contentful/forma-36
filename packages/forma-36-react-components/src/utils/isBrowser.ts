const isBrowser = () => typeof window !== 'undefined';

export default isBrowser;
