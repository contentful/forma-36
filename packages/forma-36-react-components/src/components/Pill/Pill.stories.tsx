import React from 'react';
import { action } from '@storybook/addon-actions';
import Icon from '../Icon';

import Pill, { PillProps } from './Pill';
import notes from './README.md';

export default {
  title: 'Components/Pill',
  component: Pill,
  parameters: {
    propTypes: [Pill['__docgenInfo']],
    notes,
  },
  argTypes: {
    label: { control: { type: 'text' } },
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
    onClose: { control: { disable: true } },
    onDrag: { control: { disable: true } },
    dragHandleComponent: { control: { disable: true } },
  },
};

export const basic = (args: PillProps) => <Pill label={args.label} />;

export const onDragAndOnClose = (args: PillProps) => (
  <>
    <div>
      <Pill label={args.label} onClose={args.onClose} onDrag={args.onDrag} />
    </div>
    <div className="f36-margin-top--m">
      <Pill
        style={{ width: 200 }}
        label={args.label}
        onClose={args.onClose}
        onDrag={args.onDrag}
      />
    </div>
  </>
);

export const CustomHandleComponent = (args: PillProps) => (
  <Pill
    label={args.label}
    onDrag={args.onDrag}
    dragHandleComponent={
      <Icon
        icon="ThumbUp"
        color="primary"
        style={{ padding: '0.375rem 0.625rem' }}
      />
    }
  />
);

basic.args = { label: 'example.user@contentful.com' };

onDragAndOnClose.args = {
  label: 'example.user@contentful.com ',
  onClose: action('clicked'),
  onDrag: action('dragged'),
};

CustomHandleComponent.args = {
  label: 'example.user@contentful.com',
  onClose: action('clicked'),
  onDrag: action('dragged'),
};
