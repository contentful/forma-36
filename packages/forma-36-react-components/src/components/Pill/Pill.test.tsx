import React from 'react';
import { render } from '@testing-library/react';

import { axe } from '../../utils/axeHelper';
import { Pill } from './Pill';

it('renders the component', () => {
  const { container } = render(<Pill label="test" />);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a dragging handle', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" onDrag={() => {}} />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a close button', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" onClose={() => {}} />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component in "idle" variant by default', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component in "active" variant', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" variant="active" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component in "deleted" variant', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" variant="deleted" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders no close button in "deleted" variant', () => {
  const { container } = render(
    <Pill
      className="my-extra-class"
      label="test"
      variant="deleted"
      onClose={() => {}}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with a test id', () => {
  const { container } = render(
    <Pill className="my-extra-class" label="test" testId="test-id" />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(<Pill label="test" />);
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
