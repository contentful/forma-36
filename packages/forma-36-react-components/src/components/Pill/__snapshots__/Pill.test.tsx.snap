// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div
  class="Pill"
  data-test-id="cf-ui-pill"
>
  <span
    aria-label="test"
    class="Pill__label"
    title="test"
  >
    test
  </span>
</div>
`;

exports[`renders the component with a close button 1`] = `
<div
  class="Pill my-extra-class"
  data-test-id="cf-ui-pill"
>
  <span
    aria-label="test"
    class="Pill__label"
    title="test"
  >
    test
  </span>
  <button
    aria-label="close"
    class="Pill__close-button"
    type="button"
  >
    <span
      class="TabFocusTrap"
      tabindex="-1"
    >
      <svg
        class="Icon Icon--small Icon--muted Pill__icon"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`renders the component with a dragging handle 1`] = `
<div
  class="Pill my-extra-class"
  data-test-id="cf-ui-pill"
>
  <span
    class="Pill__drag-icon"
  >
    <svg
      class="Icon Icon--small Icon--muted Pill__icon"
      data-test-id="cf-ui-icon"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </span>
  <span
    aria-label="test"
    class="Pill__label"
    title="test"
  >
    test
  </span>
</div>
`;

exports[`renders the component with a test id 1`] = `
<div
  class="Pill my-extra-class"
  data-test-id="test-id"
>
  <span
    aria-label="test"
    class="Pill__label"
    title="test"
  >
    test
  </span>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="Pill my-extra-class"
  data-test-id="cf-ui-pill"
>
  <span
    aria-label="test"
    class="Pill__label"
    title="test"
  >
    test
  </span>
</div>
`;
